File: BrainExtraction/Infant/T1T2/trainModel.R
library( ANTsR )
library( ANTsRNet )
library( keras )
library( tensorflow )
library( ggplot2 )

keras::backend()$clear_session()

Sys.setenv( "CUDA_VISIBLE_DEVICES" = "0" )

baseDirectory <- '/home/ntustison/Data/InfantMaskData/'
scriptsDirectory <- paste0( baseDirectory, 'Scripts/' )
source( paste0( scriptsDirectory, 'batchGenerator.R' ) )

templateDirectory <- paste0( baseDirectory, '../Templates/Kirby/SymmetricTemplate/' )
template <- antsImageRead( paste0( templateDirectory, "S_template3_resampled2.nii.gz" ) ) %>% iMath( "Normalize" )
templateSize <- dim( template )

################################################
#
#  Create the model and load weights
#
################################################

classes <- c( "background", "brain" )
numberOfClassificationLabels <- length( classes )
imageModalities <- c( "T1", "T2" )
channelSize <- length( imageModalities )

unetModel <- createUnetModel3D( c( templateSize, channelSize ),
  numberOfOutputs = numberOfClassificationLabels, mode = 'classification',
  numberOfLayers = 4, numberOfFiltersAtBaseLayer = 8, dropoutRate = 0.0,
  convolutionKernelSize = c( 3, 3, 3 ), deconvolutionKernelSize = c( 2, 2, 2 ),
  weightDecay = 1e-5 )

brainWeightsFileName <- paste0( scriptsDirectory, "/brainExtractionInfant.h5" )
if( file.exists( brainWeightsFileName ) )
  {
  load_model_weights_hdf5( unetModel, filepath = brainWeightsFileName )
  # } else {
  # stop( "Weights file doesn't exist.\n" )  
  }
unetModel %>% compile(
  optimizer = optimizer_adam(),
  loss = "categorical_crossentropy",
  metrics = 'accuracy' )

################################################
#
#  Load the brain data
#
################################################

cat( "Loading brain data.\n" )

brainBaseDataDirectory <- paste0( baseDirectory, "Nifti/" )
maskImages <- list.files( path = brainBaseDataDirectory, pattern = "mask.nii.gz", recursive = TRUE, full.names = TRUE )

trainingImageFiles <- list()
trainingBrainMaskFiles <- c()

pb <- txtProgressBar( min = 0, max = length( maskImages ), style = 3 )

count <- 1
for( i in seq_len( length( maskImages ) ) )
  {
  setTxtProgressBar( pb, i )

  mask <- maskImages[i]
  t1 <- gsub( "mask", "T1", mask )
  t2 <- gsub( "mask", "T2", mask )

  if( ! file.exists( t1 ) || ! file.exists( t2 ) || ! file.exists( mask ) )
    {
    next  
    # stop( "Mask doesn't exist." )  
    }

  # Do a quick check  
  # maskImage <- antsImageRead( mask )
  # if( sum( maskImage ) < 1000 )
  #   {
  #   cat( "mask: ", mask, "\n" )  
  #   next  
  #   }
  # imageArray <- as.array( antsImageRead( t1 ) )
  # if( any( is.na( imageArray ) ) || ( sum( imageArray ) <= 1000 ) )
  #   {
  #   cat( "image: ", t1, "\n" )  
  #   next  
  #   }
  # imageArray <- as.array( antsImageRead( t2 ) )
  # if( any( is.na( imageArray ) ) || ( sum( imageArray ) <= 1000 ) )
  #   {
  #   cat( "image: ", t2, "\n" )  
  #   next  
  #   }
    
  trainingImageFiles[[count]] <- c( t1, t2 )
  trainingBrainMaskFiles <- append( trainingBrainMaskFiles, mask )
  count <- count + 1
  }
cat( "\n" )  

cat( "Total training image files: ", length( trainingImageFiles ), "\n" )

cat( "\nTraining\n\n" )

###
#
# Set up the training generator
#

batchSize <- 8L

# Split trainingData into "training" and "validation" componets for
# training the model.

numberOfData <- length( trainingImageFiles )
sampleIndices <- sample( numberOfData )

validationSplit <- floor( 0.8 * numberOfData )
trainingIndices <- sampleIndices[1:validationSplit]
numberOfTrainingData <- length( trainingIndices )
validationIndices <- sampleIndices[( validationSplit + 1 ):numberOfData]
numberOfValidationData <- length( validationIndices )

###
#
# Run training
#

track <- unetModel %>% fit_generator(
  generator = batchGenerator( batchSize = batchSize,
                              imageSize = templateSize, 
                              template = template,
                              images = trainingImageFiles[trainingIndices],
                              brainMasks = trainingBrainMaskFiles[trainingIndices],
                              doRandomContralateralFlips = FALSE,
                              doDataAugmentation = TRUE
                            ),
  steps_per_epoch = 100L,
  epochs = 200L,
  validation_data = batchGenerator( batchSize = batchSize,
                                    imageSize = templateSize, 
                                    template = template,
                                    images = trainingImageFiles[validationIndices],
                                    brainMasks = trainingBrainMaskFiles[validationIndices],
                                    doRandomContralateralFlips = FALSE,
                                    doDataAugmentation = TRUE
                                  ),  
  validation_steps = 40L,
  callbacks = list(
    callback_model_checkpoint( brainWeightsFileName,
      monitor = 'val_loss', save_best_only = TRUE, save_weights_only = TRUE,
      verbose = 1, mode = 'auto' ), 
     callback_reduce_lr_on_plateau( monitor = 'val_loss', factor = 0.3,
       verbose = 1, patience = 10, mode = 'auto' ),
    callback_early_stopping( monitor = 'val_loss', min_delta = 0.001,
      patience = 20 )
  )
)

save_model_weights_hdf5( unetModel, filepath = brainWeightsFileName )

File: BrainExtraction/MRA/train_model.py
import ants
import antspynet

import os
os.environ["CUDA_VISIBLE_DEVICES"] = "1"
os.environ["ITK_DEFAULT_GLOBAL_NUMBER_OF_THREADS"] = "4"
import glob

import tensorflow as tf
import tensorflow.keras as keras
import tensorflow.keras.backend as K

from batch_generator import batch_generator

K.clear_session()
gpus = tf.config.experimental.list_physical_devices("GPU")
if len(gpus) > 0:
    tf.config.experimental.set_memory_growth(gpus[0], True)

tf.compat.v1.disable_eager_execution()
base_directory = '/home/ntustison/Data/MRA/BrainExtraction/'
scripts_directory = base_directory

from batch_generator import batch_generator

template = ants.image_read(antspynet.get_antsxnet_data("S_template3"))
template_brain_mask = antspynet.brain_extraction(template, modality="t1")

template_size = template.shape

################################################
#
#  Create the model and load weights
#
################################################

classes = ['background', 'brain']
number_of_classification_labels = len(classes)
image_modalities = ["T1"]
channel_size = len(image_modalities)

binary_dice_loss = antspynet.binary_dice_coefficient(smoothing_factor=0.0)
surface_loss = antspynet.binary_surface_loss()
ce_loss = tf.keras.losses.BinaryCrossentropy(from_logits=False)

unet_model = antspynet.create_unet_model_3d((*template_size, channel_size),
    mode="sigmoid",
    number_of_outputs=1,
    number_of_filters=(16, 32, 64, 128), dropout_rate=0.0,
    convolution_kernel_size=3, deconvolution_kernel_size=2,
    weight_decay=1e-5)

weights_filename = scripts_directory + "brainExtractionMra.h5"
if os.path.exists(weights_filename):
    unet_model.load_weights(weights_filename)

unet_model.compile(optimizer=tf.keras.optimizers.legacy.Adam(),
                   loss=binary_dice_loss,
                   metrics=[binary_dice_loss])


################################################
#
#  Load the brain data
#
################################################

print("Loading braindata.")

base_data_directory = base_directory + '/../ITKTubeTK/TemplateNifti/'
mask_images = glob.glob(base_data_directory + "Normal*/BrainMask.nii.gz")

training_image_files = list()
training_mask_files = list()

for i in range(len(mask_images)):
    mask = mask_images[i]
    image = mask.replace("BrainMask", "MRA")

    if not os.path.exists(image) or not os.path.exists(mask):
        # print(mask + " ---> " + image)
        continue

    training_image_files.append(image)
    training_mask_files.append(mask)

base_data_directory = base_directory + '../brains/'
mask_images = glob.glob(base_data_directory + "Masks2/*.nii.gz")

for i in range(len(mask_images)):
    mask = mask_images[i]
    image = mask.replace("Masks2", "Nifti")

    if not os.path.exists(image) or not os.path.exists(mask):
        continue

    training_image_files.append(image)
    training_mask_files.append(mask)


print("Total training image files: ", len(training_image_files))
print( "Training")


###
#
# Set up the training generator
#

batch_size = 4


generator = batch_generator(batch_size=batch_size,
                            image_size=template_size,
                            template=template,
                            template_brain_mask=template_brain_mask,                            
                            images=training_image_files,
                            brain_masks=training_mask_files,
                            do_random_contralateral_flips=True,
                            do_histogram_intensity_warping=True,
                            do_simulate_bias_field=True,
                            do_add_noise=True,
                            do_data_augmentation=True)


track = unet_model.fit(x=generator, epochs=200, verbose=1, steps_per_epoch=32,
    callbacks=[
       keras.callbacks.ModelCheckpoint(weights_filename, monitor='loss',
           save_best_only=True, save_weights_only=True, mode='auto', verbose=1),
       keras.callbacks.ReduceLROnPlateau(monitor='loss', factor=0.95,
          verbose=1, patience=20, mode='auto'),
    #    keras.callbacks.EarlyStopping(monitor='loss', min_delta=0.000001,
    #       patience=20)
       ]
   )

unet_model.save_weights(weights_filename)



